### PR TITLE
[Feature] Filtered MRR metrics for R-GCN example

### PR DESCRIPTION
**Pull request by hgalioulline**
## Description
Attempts to resolve issue #1275 by adding filtered MRR metrics for the R-GCN example. 
Possible reviewers: @classicsong and @yzh119.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented (debatable)
- [x] Related issue is referred in this PR

## Changes

- [x] Filtered MRR in R-GCN example

### CODE REVIEW

File: examples/pytorch/rgcn/utils.py [hunk 1]
```diff
@@ -221,3 +220,104 @@ def calc_mrr(embedding, w, test_triplets, hits=[], eval_bz=100):
...
+    return torch.LongTensor(ranks)
+
+def perturb_s_and_get_filtered_rank(embedding, w, s, r, o, test_size, triplets_to_filter):
+    """ Perturb subject in the triplets
+    """
+    num_entities = embedding.shape[0]
+    ranks = []
+    for idx in range(test_size):
+        if idx % 100 == 0:
+            print("test triplet {} / {}".format(idx, test_size))
```
Comment (classicsong) [inline examples/pytorch/rgcn/utils.py]: You should move this for loop into calc_filtered_mrr. Otherwise there will be two print output of "test triplet {} / {}" in both perturb_s_and_get_filtered_rank and perturb_o_and_get_filtered_rank which is wired.

File: examples/pytorch/rgcn/link_predict.py [hunk 2]
```diff
@@ -186,8 +186,8 @@ def main(args):
             model.eval()
             print("start eval")
             embed = model(test_graph, test_node_id, test_rel, test_norm)
-            mrr = utils.calc_mrr(embed, model.w_relation, valid_data,
-                                 hits=[1, 3, 10], eval_bz=args.eval_batch_size)
+            mrr = utils.calc_filtered_mrr(embed, model.w_relation, torch.LongTensor(train_data),
```
[the remark this answers is not in the document]
Comment (hgalioulline) [reply examples/pytorch/rgcn/link_predict.py]: Thank you for your suggestion. Something like this?
```suggestion
            raw_mrr = utils.calc_mrr(embed, model.w_relation, test_data, hits=[1, 3, 10])
            filtered_mrr = utils.calc_filtered_mrr(embed, model.w_relation, torch.LongTensor(train_data),
                                          valid_data, test_data, hits=[1, 3, 10])
            # save best model
            if filtered_mrr < best_mrr:
                if epoch >= args.n_epochs:
                    break
            else:
                best_mrr = filtered_mrr
                torch.save({'state_dict': model.state_dict(), 'epoch': epoch},
                           model_state_file)
```
Downside here is that raw_mrr wouldn't be used and during validation both raw and filtered metrics would be computed, doubling the time of evaluation.

File: examples/pytorch/rgcn/utils.py (hunk 1 above)
Comment (hgalioulline) [reply examples/pytorch/rgcn/utils.py]: Thank you for your suggestion. I would disagree that this is weird, since one print statement tracks the progress of perturbing the objects, and the other tracks the progress of perturbing subjects. This is made clear with print('Perturbing subject...') and print('Perturbing object...') in calc_filtered_mrr. The implementation of the raw MRR follows the same structure.

Comment (classicsong) [reply examples/pytorch/rgcn/utils.py]: Yep, adding "Perturbing subject .." and "Perturbing object .." is also another choice.

File: examples/pytorch/rgcn/link_predict.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply examples/pytorch/rgcn/link_predict.py]: What I mean is adding an new argument as --eval_func "mrr/filtered-mrr". We can set it filtered-mrr by default. As filtered-mrr is much more costly than mrr.
BTW, can you help profiling the seed of mrr and filtered-mrr?

[the remark this answers is not in the document]
Comment (hgalioulline) [reply examples/pytorch/rgcn/link_predict.py]: I have now added an evaluation protocol option in my latest commit :)

For the FB15k-237 dataset on my laptop, raw evaluation takes 16 minutes and filtered evaluation takes 13 minutes.

**Comment by classicsong** [comment]
LGTM

**Comment by VoVAllen** [comment]
Is there any change on performance? or updating the readme?

**Comment by classicsong** [comment]
> Is there any change on performance? or updating the readme?

@hgalioulline Can you update the Readme to add the result of filtered MRR score as well as the training scripts?

**Comment by classicsong** [comment]
@hgalioulline I have update the readme.
There speed of filterred mrr seems OK.
Thank you for your great contribution !
